pkg/karmadactl/exec.go

```diff
@@ -93,292 +86,47 @@ func execExample(parentCommand string) string {
-	GetPodTimeout time.Duration
-	Config        *restclient.Config
+	// flags specific to exec
+	KubectlExecOptions *kubectlexec.ExecOptions
+	Namespace          string
+	Cluster            string
 }
 
 // Complete verifies command line arguments and loads data from the command environment
-func (p *ExecOptions) Complete(karmadaConfig KarmadaConfig, argsIn []string, argsLenAtDash int) error {
-	if len(argsIn) > 0 && argsLenAtDash != 0 {
-		p.ResourceName = argsIn[0]
-	}
-	if argsLenAtDash > -1 {
-		p.Command = argsIn[argsLenAtDash:]
-	} else if len(argsIn) > 1 {
-		if !p.Quiet {
-			fmt.Fprint(p.ErrOut, "kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.\n")
-		}
-		p.Command = argsIn[1:]
-	} else if len(argsIn) > 0 && len(p.FilenameOptions.Filenames) != 0 {
-		if !p.Quiet {
-			fmt.Fprint(p.ErrOut, "kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.\n")
-		}
-		p.Command = argsIn[0:]
-		p.ResourceName = ""
-	}
-
-	var err error
-
-	if len(p.Cluster) == 0 {
-		return fmt.Errorf("must specify a cluster")
-	}
-
-	karmadaRestConfig, err := karmadaConfig.GetRestConfig(p.KarmadaContext, p.KubeConfig)
+func (o *ExecOptions) Complete(karmadaConfig KarmadaConfig, cmd *cobra.Command, argsIn []string, argsLenAtDash int) error {
+	karmadaRestConfig, err := karmadaConfig.GetRestConfig(o.KarmadaContext, o.KubeConfig)
 	if err != nil {
 		return fmt.Errorf("failed to get control plane rest config. context: %s, kube-config: %s, error: %v",
-			p.KarmadaContext, p.KubeConfig, err)
+			o.KarmadaContext, o.KubeConfig, err)
 	}
-
-	clusterInfo, err := getClusterInfo(karmadaRestConfig, p.Cluster, p.KubeConfig, p.KarmadaContext)
+	clusterInfo, err := getClusterInfo(karmadaRestConfig, o.Cluster, o.KubeConfig, o.KarmadaContext)
 	if err != nil {
 		return err
 	}
-
-	f := getFactory(p.Cluster, clusterInfo)
-
-	p.Namespace, p.EnforceNamespace, err = f.ToRawKubeConfigLoader().Namespace()
-	if err != nil {
-		return err
-	}
-
-	p.ExecutablePodFn = polymorphichelpers.AttachablePodForObjectFn
-
-	p.Builder = f.NewBuilder
-	p.restClientGetter = f
-
-	p.Config, err = f.ToRESTConfig()
-	if err != nil {
-		return err
-	}
-
-	clientset, err := f.KubernetesClientSet()
-	if err != nil {
-		return err
-	}
-	p.PodClient = clientset.CoreV1()
-
-	return nil
+	// Build member cluster kubeConfigFlags
+	kubeConfigFlags := NewConfigFlags(true).WithDeprecatedPasswordFlag()
+	kubeConfigFlags.APIServer = stringptr(clusterInfo[o.Cluster].APIEndpoint)
+	kubeConfigFlags.KubeConfig = stringptr(clusterInfo[o.Cluster].KubeConfig)
+	kubeConfigFlags.Context = stringptr(clusterInfo[o.Cluster].Context)
+	kubeConfigFlags.usePersistentConfig = true
+	if o.Namespace != "" {
+		kubeConfigFlags.Namespace = &o.Namespace
+	}
+	matchVersionKubeConfigFlags := cmdutil.NewMatchVersionFlags(kubeConfigFlags)
+	f := cmdutil.NewFactory(matchVersionKubeConfigFlags)
```

Comment:
I think we can refactor function `getFactory(p.Cluster, clusterInfo)` to `getFactory(p.Cluster, clusterInfo, namespace)` that also used by `logs`, `get`, `describe`.

Reply:
It's make sense. I'll do it in another pull request because we must change a lot of files after refactoring getFactory function, it has gone beyond the focus of this commit.

@lonelyCZ  what do you think?